test/test.hs

```diff
@@ -70,6 +92,336 @@ main = defaultMain tests
+                              Lit a -> a
+                              _ -> error "Timestamp needs to be a Lit"
+
+emptyEvmToolEnv :: EVMToolEnv
+emptyEvmToolEnv = EVMToolEnv { _coinbase = 0
+                             , _timestamp = Lit 0
+                             , _number     = 0
+                             , _prevRandao = 42069
+                             , _gasLimit   = 0xffffffffffffffff
+                             , _baseFee    = 0
+                             , _maxCodeSize= 0xffffffff
+                             , _schedule   = FeeSchedule.berlin
+                             , _blockHashes = mempty
+                             }
+
+data EVMToolReceipt =
+  EVMToolReceipt
+    { recType :: String
+    , recRoot :: String
+    , recStatus :: String
+    , recCumulativeGasUsed :: String
+    , recLogsBloom :: String
+    , recLogs :: Maybe String
+    , recTransactionHash :: String
+    , recContractAddress :: String
+    , recGasUsed :: String
+    , recBlockHash :: String
+    , recTransactionIndex :: String
+    } deriving (Generic, Show)
+
+instance JSON.FromJSON EVMToolReceipt where
+    parseJSON = JSON.withObject "EVMReceipt" $ \v -> EVMToolReceipt
+        <$> v .: "type"
+        <*> v .: "root"
+        <*> v .: "status"
+        <*> v .: "cumulativeGasUsed"
+        <*> v .: "logsBloom"
+        <*> v .: "logs"
+        <*> v .: "transactionHash"
+        <*> v .: "contractAddress"
+        <*> v .: "gasUsed"
+        <*> v .: "blockHash"
+        <*> v .: "transactionIndex"
+
+data EVMToolResult =
+  EVMToolResult
+  { stateRoot :: String
+  , txRoot :: String
+  , receiptsRoot :: String
+  , logsHash :: String
+  , logsBloom :: String
+  , receipts :: [EVMToolReceipt]
+  , currentDifficulty :: String
+  , gasUsed :: String
+  , rejected :: Maybe [EVMRejected]
+  } deriving (Generic, Show)
+
+instance JSON.FromJSON EVMToolResult
+
+data EVMRejected =
+  EVMRejected
+    { index :: Int
+    , err :: String
+    } deriving (Generic, Show)
+
+instance JSON.FromJSON EVMRejected where
+  parseJSON = JSON.withObject "EVMRejected" $ \v -> EVMRejected
+    <$> v .: "index"
+    <*> v .: "error"
+
+data EVMToolAlloc =
+  EVMToolAlloc
+  { balance :: W256 -- this is wrongly written by default, hence the non-default ToJSON
+  , code :: ByteString
+  , nonce :: W64
+  } deriving (Generic)
+
+instance JSON.ToJSON EVMToolAlloc where
+  toJSON b = JSON.object [ ("balance" ,  (JSON.toJSON $ show b.balance))
+                         , ("code", (JSON.toJSON $ b.code))
+                         , ("nonce"  , (JSON.toJSON $ b.nonce))
+                         ]
+
+emptyEVMToolAlloc :: EVMToolAlloc
+emptyEVMToolAlloc = EVMToolAlloc { balance = 0
+                                 , code = mempty
+                                 , nonce = 0
+                                 }
+
+getEVMToolRet :: OpContract -> ByteString -> Int -> IO (Maybe EVMToolResult, (Either (EVM.Error, [VMTrace]) (Expr 'End, [VMTrace], VMTraceResult)))
```

Comment:
this function seems to be doing more than it's name suggests, and the return type is pretty hard to parse. perhaps this could be tidied a little?

Reply:
Thanks, yeah, it was a huge mess. I now cut it into two functions. Should be a lot cleaner :)

Reply:
Let me know what you think of the new setup!